test/cpp_api_parity/functional_impl_check.py

```diff
@@ -0,0 +1,232 @@
+# ${cpp_function_call}
+TORCH_NN_FUNCTIONAL_TEST_FORWARD = Template("""
+void ${functional_variant_name}_test_forward(
+    const std::string& arg_dict_file_path,
+    const std::string& forward_output_file_path) {
+  pybind11::gil_scoped_release no_gil;
+
+  namespace F = torch::nn::functional;
+
+  // Declare arguments
+  auto arg_dict = load_dict_from_file(arg_dict_file_path);
+  ${cpp_args_construction_stmts};
+
+  // Some functionals (such as `F::rrelu`) create random tensors in their call path.
+  // To make sure the random tensors created are the same in Python/C++, we need
+  // to set the RNG seed manually.
+  torch::manual_seed(0);
+
+  // Run function with arguments
+  auto cpp_output = ${cpp_function_call};
+
+  // Save the output into a file to be compared in Python later
+  write_ivalue_to_file(torch::IValue(cpp_output), forward_output_file_path);
+}
+""")
+
+def run_forward(unit_test_class, test_params):
+    device = test_params.device
+
+    inputs = set_python_tensors_requires_grad([arg_value for _, arg_value in test_params.arg_dict['input']])
+    inputs = inputs + [arg_value for _, arg_value in test_params.arg_dict['target']]
+    inputs = inputs + [arg_value for _, arg_value in test_params.arg_dict['extra_args']]
+    inputs = move_python_tensors_to_device(inputs, device)
+
+    # Some functionals (such as `F.rrelu`) create random tensors in their call path.
+    # To make sure the random tensors created are the same in Python/C++, we need
+    # to set the RNG seed manually.
+    torch.manual_seed(0)
+    python_output = test_params.test_instance.constructor()(*inputs)
+
+    return python_output
+
+def test_forward(unit_test_class, test_params):
+    functional_variant_name = test_params.functional_variant_name
+    cpp_tmp_folder = test_params.cpp_tmp_folder
+
+    # Run forward on Python functional
+    python_output = run_forward(unit_test_class, test_params)
+
+    # Save Python arguments to be used from C++ function
+    arg_dict_file_path = compute_temp_file_path(cpp_tmp_folder, functional_variant_name, 'arg_dict')
+    serialize_arg_dict_as_script_module(test_params.arg_dict).save(arg_dict_file_path)
+
+    cpp_test_name = '{}_{}'.format(test_params.functional_variant_name, 'test_forward')
+    cpp_test_fn = getattr(unit_test_class.functional_impl_check_cpp_module, cpp_test_name)
+
+    def run_cpp_test_fn_and_check_output():
+        forward_output_file_path = compute_temp_file_path(cpp_tmp_folder, functional_variant_name, 'forward_output')
+
+        cpp_test_fn(arg_dict_file_path, forward_output_file_path)
+        cpp_output = torch.load(forward_output_file_path)
+
+        # Check that forward outputs are equal
+        unit_test_class.assertTrue(
+            torch.allclose(python_output, cpp_output),
+            generate_error_msg("forward output", cpp_output, python_output))
+
+    run_cpp_test_fn_and_check_output()
+
+    # Remove temporary folder that stores C++ outputs
+    shutil.rmtree(cpp_tmp_folder)
```

Comment:
This will raise an exception on error, should we add a comments here say this will raise exception on error?  or catch it?

Reply:
Thanks for the catch! I added an assert above this line to check that `cpp_tmp_folder` always exists on disk (since the current form of the test mechanism does expect it to exist) :D

Reply:
What I am concern is that since this is to delete from hdd/sdd file system, it is not always reliable depending on how many instances currently using that machine and accessing the same disk, if your deletion request failed due to high disk traffic or some other temp connection issues, this shouldn't block the whole process. But I think facebook's cluster is well scaled generally, above issue should be catch in upper hardware layer, adding an assert here  to check folder existence is fine then.

Reply:
@glaringlee we run these tests on OSS as well (I'm not sure how many of these tests get run internally). If we're really scared of flakiness from `rmtree` not working, we could wrap the statement in a try block and don't catch any exceptions (lol):
```
try:
    shutil.rmtree(cpp_tmp_folder)
```

Reply:
@zou3519 @yf225, I am not sure how often this would happen, I do have experience that such failure happened one to two times a month on average. I agree that a single try would work just to make sure this wouldn't block the process even failed.

Reply:
Thanks for the discussion! I added
```python
try:
    shutil.rmtree(cpp_tmp_folder)
except:
    pass
```
because otherwise Python is complaining about wrong indentation level at next line..